lib/commons/color/get-text-shadow-colors.js

```diff
@@ -22,29 +31,96 @@ function getTextShadowColors(node, { minRatio, maxRatio } = {}) {
+  const thinShadows = [];
   const shadows = parseTextShadows(textShadow);
-  shadows.forEach(({ colorStr, pixels }) => {
+  for (const shadow of shadows) {
     // Defaults only necessary for IE
-    colorStr = colorStr || style.getPropertyValue('color');
-    const [offsetY, offsetX, blurRadius = 0] = pixels;
-    if (
-      (!minRatio || blurRadius >= fontSize * minRatio) &&
-      (!maxRatio || blurRadius < fontSize * maxRatio)
-    ) {
-      const color = textShadowColor({
-        colorStr,
-        offsetY,
-        offsetX,
-        blurRadius,
-        fontSize
-      });
-      shadowColors.push(color);
+    const colorStr = shadow.colorStr || style.getPropertyValue('color');
+    const [offsetX, offsetY, blurRadius = 0] = shadow.pixels;
+    if (maxRatio && blurRadius >= fontSize * maxRatio) {
+      continue;
     }
-  });
+    if (minRatio && blurRadius < fontSize * minRatio) {
+      thinShadows.push({ colorStr, pixels: shadow.pixels });
+      continue;
+    }
+    if (thinShadows.length > 0) {
+      // Inset any stroke colors before this shadow
+      const strokeColors = getStrokeColor(thinShadows);
+      shadowColors.push(...strokeColors);
+      thinShadows.splice(0, thinShadows.length); // empty
+    }
+
+    const color = textShadowColor({
+      colorStr,
+      offsetX,
+      offsetY,
+      blurRadius,
+      fontSize
+    });
+    shadowColors.push(color);
+  }
+
+  if (thinShadows.length > 0) {
+    // Append any remaining stroke colors
+    const strokeColors = getStrokeColor(thinShadows);
+    shadowColors.push(...strokeColors);
+  }
+
   return shadowColors;
 }
 
+function getStrokeColor(thinShadows) {
+  const edges = ['top', 'right', 'bottom', 'left'];
+  const colorMap = {};
+  for (const { colorStr, pixels } of thinShadows) {
+    colorMap[colorStr] ??= { top: [], right: [], bottom: [], left: [] };
+    const borders = colorMap[colorStr];
+    const [offsetX, offsetY] = pixels;
+
+    if (offsetX > VISIBLE_SHADOW_MIN_PX) {
+      borders.right.push(pixels);
+    } else if (offsetX < -VISIBLE_SHADOW_MIN_PX) {
+      borders.left.push(pixels);
+    }
+    if (offsetY > VISIBLE_SHADOW_MIN_PX) {
+      borders.bottom.push(pixels);
+    } else if (offsetY < -VISIBLE_SHADOW_MIN_PX) {
+      borders.top.push(pixels);
+    }
+  }
+
+  const strokeShadows = Object.entries(colorMap).filter(([, sides]) => {
+    const sidesCovered = edges.filter(side => sides[side].length !== 0).length;
+    return sidesCovered >= 3;
+  });
+
+  const strokeColors = strokeShadows.map(([colorStr, sides]) => {
+    const strokeColor = new Color();
+    strokeColor.parseString(colorStr);
+
+    // Average the number of shadows around the text
+    let density = 0;
+    let isSolid = true;
+    edges.forEach(edge => {
+      density += sides[edge].length / 4;
+      isSolid &&= sides[edge].every(
+        ([x, y]) =>
+          Math.abs(x) > OPAQUE_STROKE_OFFSET_MIN_PX ||
+          Math.abs(y) > OPAQUE_STROKE_OFFSET_MIN_PX
+      );
```

Comment:
This is assuming that a 1.4px shadow and a 0.6px shadow should contribute equally to density, but I don't think that's accurate in practice based on playing with the `--thickness` CSS var in [this test playground](https://developer.mozilla.org/en-US/play?id=b4w6cdYxibzUleKHyvYH%2BRXJbVLocqnjdC82xafXh21t4sfT4M%2FM%2F24pjp5%2BvuaSJm7TToYJTgsBuSrq). Perhaps better to weight a side's contribution to density according to its relevant x/y value, as opposed to each side contributing a flat 1? I think the below covers that such that the behavior is the same as the original magic algorithm for 1px thick shadows

```suggestion
      for (const side of sides[edge]) {
        const relevantSideProperty = { left: 'x', right: 'x', top: 'y', bottom: 'y' }[edge];
        const sideThickness = Math.abs(side[relevantSideProperty])
        isSolid &&= sideThickness > OPAQUE_STROKE_OFFSET_MIN_PX
        density += Math.min(sideThickness, OPAQUE_STROKE_OFFSET_MIN_PX)
      }
```

Reply:
The reason I did that is because I noticed when you want half-pixels shadows aren't distributed evenly. When I did a .6px I got 70% alpha on one side of the text and 16% on the opposite side. On the other hand we do sort of still pull an average, so this makes sense.

Reply:
That's a convincing enough reason to give these cases uniform density. Maybe leave a comment explaining that reasoning?

I do think the functional change I've made here to `isSolid` would be good to pull in regardless, though; I think the original logic would treat a `.5px 2px #000, -2px -.5px #000` shadow as solid when it doesn't intend to. Maybe a cleaner fix for that would be to update further up when you're building up `colorMap` (the 4 lines like `borders.N.push(pixels)`) to push only the relevant `offsetN` into the side's array instead of `pixels`?